### PR TITLE
LOK : 2449 : Alerts: NodeInfo should be deleted when a node is deleted

### PR DESCRIPTION
**Pull request by Junaid-Khan-Nant**
## Description
- Currently we only send Node related updates to Alert Service when we save/update a Node. We should also send an update to remove nodes when a node gets deleted.
- We can do this with an operation to update/remove similar to  TagOperationProto.
- We need to update NodeKafkaProducer to send updates to Kafka even when we delete node.
- Alert Service should consume that update and delete Node from it’s table.

## Jira link(s)
https://opennms.atlassian.net/browse/LOK-2449

## Checklist
* [ ] Follows Lōkahi's [development guidelines.](https://github.com/OpenNMS-Cloud/lokahi/wiki/Development-Guidelines)
* [ ] Appropriate reviewer(s) have been selected.
* [ ] Jira issue(s) have been updated to "In Review".
* [ ] Includes [appropriate tests.](https://github.com/OpenNMS-Cloud/lokahi/wiki/Test-Strategy)
* [ ] Documentation has been updated as necessary.
* [ ] Notify devops of changes to the Charts.
* [ ] Notify documentation team of any changes to names of screens or features (affects URLs).

### CODE REVIEW

File: alert/src/main/java/org/opennms/horizon/alertservice/api/AlertService.java [hunk 1]
```diff
@@ -50,4 +51,6 @@ public interface AlertService {
     void saveNode(NodeDTO node);
 
     AlertCount getAlertsCount(String tenantId);
+
+    void deleteNodeInAlert(NodeOperationProto alertOperationProto);
```
Comment (smunir-onms) [inline alert/src/main/java/org/opennms/horizon/alertservice/api/AlertService.java:55]: minor naming:  void deleteNodeInAlert(NodeOperationProto nodeOperationProto);

File: inventory/main/src/main/java/org/opennms/horizon/inventory/component/NodeKafkaProducer.java [hunk 2]
```diff
@@ -51,12 +53,22 @@ public class NodeKafkaProducer {
...
+        var producerRecord = new ProducerRecord<String, byte[]>(topic, nodeOperationProto.toByteArray());
         kafkaTemplate.send(producerRecord);
     }
 
-    public void updateNodeInKafka(NodeDTO nodeDTO) {
-        var producerRecord = new ProducerRecord<String, byte[]>(topic, nodeDTO.toByteArray());
+    public void deleteNodeInKafka(NodeOperationProto nodeOperationProto) {
+        var producerRecord = new ProducerRecord<String, byte[]>(topic, nodeOperationProto.toByteArray());
         kafkaTemplate.send(producerRecord);
     }
```
Comment (smunir-onms) [inline inventory/main/src/main/java/org/opennms/horizon/inventory/component/NodeKafkaProducer.java:73]: suggestion:
```
public void sendNode(Node node) {
        NodeDTO nodeDTO = nodeMapper.modelToDTO(node);
        NodeOperationProto nodeUpdateOperation = NodeOperationProto.newBuilder()
                .setOperation(NodeOperation.UPDATE_NODE)
                .setNodeDto(nodeDTO)
                .build();

        nodeOperationInKafka(nodeUpdateOperation);
    }

    public void nodeOperationInKafka(NodeOperationProto nodeOperationProto) {
        var producerRecord = new ProducerRecord<String, byte[]>(topic, nodeOperationProto.toByteArray());
        kafkaTemplate.send(producerRecord);
    }
```

File: inventory/main/src/main/java/org/opennms/horizon/inventory/service/NodeService.java [hunk 3]
```diff
@@ -439,7 +442,24 @@ public void updateNodeInKafka(long nodeId, String tenantId) {
         var optionalNode = getByIdAndTenantId(nodeId, tenantId);
         if (optionalNode.isPresent()) {
             var nodeDTO = optionalNode.get();
-            nodeKafkaProducer.updateNodeInKafka(nodeDTO);
+            nodeKafkaProducer.updateNodeInKafka(NodeOperationProto.newBuilder()
```
Comment (smunir-onms) [inline inventory/main/src/main/java/org/opennms/horizon/inventory/service/NodeService.java:445]: reference change as per above comment

File: inventory/main/src/main/java/org/opennms/horizon/inventory/service/NodeService.java [hunk 4]
```diff
@@ -439,7 +442,24 @@ public void updateNodeInKafka(long nodeId, String tenantId) {
...
+                    .setNodeDto(nodeDTO)
+                    .setOperation(NodeOperation.UPDATE_NODE)
+                    .build());
+        }
+    }
+
+    public void deleteNodeInKafka(Node node) {
+        var optionalNode = getByIdAndTenantId(node.getId(), node.getTenantId());
+        if (optionalNode.isPresent()) {
+            var val = NodeOperationProto.newBuilder()
```
Comment (smunir-onms) [inline inventory/main/src/main/java/org/opennms/horizon/inventory/service/NodeService.java:455]: var nodeDeleteOperation

reference change: 
nodeKafkaProducer.nodeOperationInKafka(nodeDeleteOperation);

File: inventory/main/src/test/java/org/opennms/horizon/inventory/component/NodeKafkaProducerTest.java [hunk 5]
```diff
@@ -86,13 +86,23 @@ void testSendNode() {
 
     private boolean nodeProducerRecordMatches(Node expectedNode, ProducerRecord<String, byte[]> producerRecord) {
         try {
-            NodeDTO nodeDTO = NodeDTO.parseFrom(producerRecord.value());
+            // NodeDTO nodeDTO = NodeDTO.parseFrom(producerRecord.value());
```
Comment (smunir-onms) [inline inventory/main/src/test/java/org/opennms/horizon/inventory/component/NodeKafkaProducerTest.java:89]: remove commented code

**Review by smunir-onms — COMMENTED** [review]
LGTM

File: inventory/main/src/main/java/org/opennms/horizon/inventory/service/NodeService.java [hunk 6]
```diff
@@ -439,7 +442,24 @@ public void updateNodeInKafka(long nodeId, String tenantId) {
...
-            nodeKafkaProducer.updateNodeInKafka(nodeDTO);
+            nodeKafkaProducer.nodeOperationInKafka(NodeOperationProto.newBuilder()
+                    .setNodeDto(nodeDTO)
+                    .setOperation(NodeOperation.UPDATE_NODE)
+                    .build());
+        }
+    }
+
+    public void deleteNodeInKafka(Node node) {
+        var optionalNode = getByIdAndTenantId(node.getId(), node.getTenantId());
```
Comment (cgorantla) [inline inventory/main/src/main/java/org/opennms/horizon/inventory/service/NodeService.java:453]: you should not invoke a method with @Transactional from a method without @Transactional.

see https://sonarcloud.io/project/issues?resolved=false&severities=BLOCKER%2CCRITICAL%2CMAJOR%2CMINOR&sinceLeakPeriod=true&types=BUG&pullRequest=1805&id=opennms_lokahi_inventory&open=AY7rq4PudReoRgpnugif&tab=code

**Review by cgorantla — COMMENTED** [review]
I don't see any update in alert.node table when node is created, so something is broken with this change

**Review by cgorantla — APPROVED** [review]
LGTM !

**Pull request merged**